tools/hf2megads_weight_converter.py

```diff
@@ -57,85 +71,111 @@ def compute_partition_range(hidden_size, local_rank, tp_size):
+    from transformers import AutoModelForCausalLM
     from transformers.modeling_utils import no_init_weights
 
-    if no_init:
-        hf_config = AutoConfig.from_pretrained(hf_ckpt_dir, trust_remote_code=True)
-        with no_init_weights():
-            hf_model = AutoModelForCausalLM.from_config(hf_config, trust_remote_code=True, torch_dtype=torch.bfloat16)
-    else:
-        hf_model = {}
-        hf_auto_model = AutoModelForCausalLM.from_pretrained(hf_ckpt_dir, trust_remote_code=True, torch_dtype=torch.bfloat16)
-        log.info(
-            f"----------------------------hf weight list----------------------------")
+    ctx = nullcontext() if no_init else no_init_weights()
+    with ctx:
+        hf_auto_model = AutoModelForCausalLM.from_pretrained(
+            hf_ckpt_dir, trust_remote_code=True, torch_dtype=torch.bfloat16
+        )
+    logger.info(
+        '----------------------------hf weight list----------------------------'
+    )
 
-        for name, param in hf_auto_model.named_parameters():
-            hf_model[name] = param.clone()
-            log.info(name)
+    hf_model = {}
+    for name, submodule in hf_auto_model.named_children():
+        for pname, param in submodule.named_parameters():
+            logger.info(f'[{name}.{pname}] shape={param.shape}')
+            hf_model[f'{name}.{pname}'] = param.clone()
 
+    torch.distributed.barrier()
     return hf_model
 
 
 def print_distinct_weights(model):
-    log.info(
-        f"----------------------------mega-ds weight list----------------------------")
+    logger.info(
+        '----------------------------mega-ds weight list----------------------------'
+    )
     for pipe_rank in range(mpu.get_pipeline_model_parallel_world_size()):
         if mpu.get_pipeline_model_parallel_rank() == pipe_rank:
-            if mpu.get_data_parallel_rank() == 0 and mpu.get_tensor_model_parallel_rank(
-            ) == 0:
+            if (
+                mpu.get_data_parallel_rank() == 0
+                and mpu.get_tensor_model_parallel_rank() == 0
+            ):
                 for pname, p in model.named_parameters():
-                    print(pname)
+                    logger.info(f'{pname}, {p.shape}')
             torch.distributed.barrier()
         else:
             torch.distributed.barrier()
 
 
+def get_hf_tokenizer(name: str):
+    from transformers import AutoTokenizer
+
+    tokenizer = AutoTokenizer.from_pretrained(name)
+    return tokenizer
+
+
 class refactor:
     def __init__(self, ds_model, hf_model, args, config):
-        tokenizer = get_tokenizer()
+        self.tokenizer = get_tokenizer()
+        if args.tokenizer_type == 'HFTokenizer':
+            self.hf_tokenizer = get_hf_tokenizer(args.tokenizer_model)
+            self.token_vocab = len(self.hf_tokenizer)
+        else:
+            self.hf_tokenizer = None
+            assert self.tokenizer is not None
+            self.token_vocab = self.tokenizer.vocab_size
```

Comment:
Mismatch between `self.tokenizer.vocab_size` `hf_w.shape[0]` when using Llama3 tokenizers.

This discrepancy causes the following assertion:

```python
assert hf_w.shape[0] == self.padded_vocab_size
```

to fail since `hf_w.shape[0] = 128256` but `self.tokenizer.vocab_size = 128000`.

Explicitly:

```python
>>> type(self.hf_tokenizer)
<class 'transformers.tokenization_utils_fast.PreTrainedTokenizerFast'>
>>> type(self.tokenizer)
<class 'megatron.tokenizer.tokenizer._HFTokenizer'>
>>> self.tokenizer.vocab_size
128000
>>> len(self.hf_tokenizer)
128256
```

so, replacing `self.token_vocab = len(self.hf_tokenizer)` seems to resolve this issue.